### ISSUE TITLE
oanda and yahoo sources fail with caching

### ISSUE THREAD

**Issue by blais** [beancount/beancount#287]
**[Original report](https://bitbucket.org/blais/beancount/issue/287) by Jason Chu (Bitbucket: [xentac](https://bitbucket.org/xentac), GitHub: [xentac](https://github.com/xentac)).**

----------------------------------------

If my price source looks like

```
1841-01-01 commodity CAD
  export: "CASH"
  name: "CA Dollar"
  price: "USD:yahoo/^CAD=X" 
```

or

```
1841-01-01 commodity CAD
  export: "CASH"
  name: "CA Dollar"
  price: "USD:oanda/^USD_CAD"
```

bean-price gives me this output:
```
Traceback (most recent call last):
  File "/usr/lib/python3.6/shelve.py", line 111, in __getitem__
    value = self.cache[key]
KeyError: '36af117ec1697ec0980fa6e2b9309fbf'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/bin/bean-price", line 4, in <module>
    from beancount.prices.price import main; main()
  File "/usr/lib/python3.6/site-packages/beancount/prices/price.py", line 349, in main
    price_entries = sorted(price_entries, key=lambda e: e.currency)
  File "/usr/lib/python3.6/concurrent/futures/_base.py", line 586, in result_iterator
    yield fs.pop().result()
  File "/usr/lib/python3.6/concurrent/futures/_base.py", line 425, in result
    return self.__get_result()
  File "/usr/lib/python3.6/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/usr/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/lib/python3.6/site-packages/beancount/prices/price.py", line 139, in fetch_price
    srcprice = fetch_cached_price(source, psource.symbol, dprice.date)
  File "/usr/lib/python3.6/site-packages/beancount/prices/price.py", line 84, in fetch_cached_price
    timestamp_created, result = _CACHE[key]
  File "/usr/lib/python3.6/shelve.py", line 114, in __getitem__
    value = Unpickler(f).load()
AttributeError: Can't get attribute 'tzfile' on <module 'dateutil.zoneinfo' from '/usr/lib/python3.6/site-packages/dateutil/zoneinfo/__init__.py'>
```

If I call bean-price with --no-cache, everything works awesome. From what I can gather of the error, it looks like pickling some timezones isn't supported?

**Comment by blais** [comment]
**Original comment by Vivek G (Bitbucket: [seltzered](https://bitbucket.org/seltzered), GitHub: [seltzered](https://github.com/seltzered)).**

----------------------------------------

FWIW, I'm seeing something similar,  traceback is below (using beancount src rev 4567). 


```
#!python

$bean-price -v -e "USD:yahoo/CAKE" 
INFO    : Using price cache at "/var/folders/mn/0qxnbsbx2993nhk28z3g943w0000gn/T/bean-price.cache" (with indefinite expiration)
INFO    : Processing at date: 2018-04-30
INFO    : Starting new HTTPS connection (1): query1.finance.yahoo.com
Traceback (most recent call last):
  File "/Users/me/.pyenv/versions/3.5.0/lib/python3.5/shelve.py", line 111, in __getitem__
    value = self.cache[key]
KeyError: '866176d3642c444a4cfe1e890a12d65e'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/me/projects/!clones/beancount/beancount/prices/price.py", line 84, in fetch_cached_price
    timestamp_created, result = _CACHE[key]
  File "/Users/me/.pyenv/versions/3.5.0/lib/python3.5/shelve.py", line 113, in __getitem__
    f = BytesIO(self.dict[key.encode(self.keyencoding)])
KeyError: b'866176d3642c444a4cfe1e890a12d65e'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/me/.pyenv/versions/3.5.0/bin/bean-price", line 6, in <module>
    exec(compile(open(__file__).read(), __file__, 'exec'))
  File "/Users/me/projects/!clones/beancount/bin/bean-price", line 4, in <module>
    from beancount.prices.price import main; main()
  File "/Users/me/projects/!clones/beancount/beancount/prices/price.py", line 349, in main
    price_entries = sorted(price_entries, key=lambda e: e.currency)
  File "/Users/me/.pyenv/versions/3.5.0/lib/python3.5/concurrent/futures/_base.py", line 556, in result_iterator
    yield future.result()
  File "/Users/me/.pyenv/versions/3.5.0/lib/python3.5/concurrent/futures/_base.py", line 405, in result
    return self.__get_result()
  File "/Users/me/.pyenv/versions/3.5.0/lib/python3.5/concurrent/futures/_base.py", line 357, in __get_result
    raise self._exception
  File "/Users/me/.pyenv/versions/3.5.0/lib/python3.5/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/me/projects/!clones/beancount/beancount/prices/price.py", line 139, in fetch_price
    srcprice = fetch_cached_price(source, psource.symbol, dprice.date)
  File "/Users/me/projects/!clones/beancount/beancount/prices/price.py", line 91, in fetch_cached_price
    _CACHE[key] = (timestamp_now, result)
  File "/Users/me/.pyenv/versions/3.5.0/lib/python3.5/shelve.py", line 124, in __setitem__
    p.dump(value)
_pickle.PicklingError: Can't pickle <class 'typing.SourcePrice'>: attribute lookup SourcePrice on typing failed

```

**Comment by blais** [comment]
**Original comment by Vivek G (Bitbucket: [seltzered](https://bitbucket.org/seltzered), GitHub: [seltzered](https://github.com/seltzered)).**

----------------------------------------

Updating the 'pickleshare' package from 0.6 to 0.7.4 seems to fix this issue.

It's intermittent and seem to only happen on some items. It was still happening when doing 'bean-price ledger.beancount' with pickleshare 0.7.4 on python 3.5 yet worked okay with pickleshare 0.7.4 on python 3.6.3

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

I'll make another fix that removes the timezone instead, in order to avoid a new dependency.

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

What version of Python are you guys using?
I'm on 3.6.3
I can't seem to be able to reproduce either of these problems.

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

I just installed 3.5 and I still can't reproduce.

**Comment by blais** [comment]
**Original comment by Vivek G (Bitbucket: [seltzered](https://bitbucket.org/seltzered), GitHub: [seltzered](https://github.com/seltzered)).**

----------------------------------------

I believe I was on python 3.5 / pickleshare 0.6 when I saw the issue - even then it was intermittent (e.g. it may happen on my sample of bean-price -v -e "USD:yahoo/CAKE" , but not on bean-price -v -e "USD:yahoo/TSLA"  )

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

Do you think it's possible that you had an old cache when the bug occurred?
That somehow the mix of old vs. new cache entries would make it fail?

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

I'll bet your (Vivek's) error was due to this. For Jason's error it's probably due to tz serialization.
I'll convert all timestamps to UTC then strip them to timezone naive and reverse that when I reload, in order to circumvent serialization of timezone objects.

**Comment by blais** [comment]
**Original comment by Jason Chu (Bitbucket: [xentac](https://bitbucket.org/xentac), GitHub: [xentac](https://github.com/xentac)).**

----------------------------------------

I'm using python 3.6.4.

I tried deleting the cache and running my repro and it still caused a problem for both sources.

I can do some more digging if you want, but I didn't have time when I first submitted the issue.

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

If you have a minute, can you provide a list of steps that make the problem occur? Ideally with a minimal beancount file (shave it down to the bare minimum).
Thank you,

**Comment by blais** [comment]
**Original comment by Jason Chu (Bitbucket: [xentac](https://bitbucket.org/xentac), GitHub: [xentac](https://github.com/xentac)).**

----------------------------------------

Right off the bat, I found that it was an interaction with the implicit_prices plugin.

I'll continue working on the minimal file.

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

Fixed #287: In the price fecher, store the datetime instances as naive-in-UTC in order to avoid a serialization problem with older Python versions.
